base/math.jl

```diff
@@ -95,6 +95,131 @@ macro horner(x, p...)
     return Expr(:block, :(local t = $(esc(x))), ex, :r)
 end
 
+"""
+    evalpoly(x, p::Tuple)
+Evaluate the polynomial ``\\sum_k p[k] x^{k-1}`` for the coefficients `p[1]`, `p[2]`, ...;
+that is, the coefficients are given in ascending order by power of `x`. This function
+generates efficient code using Horner's method with loops unrolled at compile time.
+# Example
+```jldoctest
+julia> evalpoly(2, (1, 2, 3))
+17
+```
+"""
+function evalpoly(x, p::Tuple)
+    if @generated
+        N = length(p.parameters)
+        ex = :(p[end])
+        for i in N-1:-1:1
+            ex = :(muladd(x, $ex, p[$i]))
+        end
+        ex
+    else
+        _evalpoly(x, p)
+    end
+end
+
+"""
+    evalpoly(x, p::AbstractVector)
+Evaluate the polynomial ``\\sum_k p[k] x^{k-1}`` for the coefficients `p[1]`, `p[2]`, ...;
+that is, the coefficients are given in ascending order by power of `x`. This function
+uses Horner's method *without* loops unrolled at compile time. Use this method when the
+number of coefficients is not known at compile time.
+# Example
+```jldoctest
+julia> evalpoly(2, [1, 2, 3])
+17
+```
+"""
+evalpoly(x, p::AbstractVector) = _evalpoly(x, p)
```

Comment:
I chose to restrict this to abstract vector, but not `_evalpoly(x, p)` (same for the complex case). I'm not sure if this is the preferred way to go about this, but it seemed like the most straightforward way to do this with minimal code duplication.